yt/utilities/lib/pixelization_routines.pyx

```diff
@@ -1278,6 +1304,93 @@ def pixelize_sph_kernel_arbitrary_grid(np.float64_t[:, :, :] buff,
                         continue
                     buff[xi, yi, zi] += buff_num[xi, yi, zi] / buff_denom[xi, yi, zi]
 
+@cython.initializedcheck(False)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def pixelize_sph_kernel_gather_arbitrary_grid(np.float64_t[:, :, :] buff,
+        np.float64_t[:] posx, np.float64_t[:] posy, np.float64_t[:] posz,
+        np.float64_t[:] hsml, np.float64_t[:] pmass, np.float64_t[:] pdens,
+        np.float64_t[:] quantity_to_smooth, np.float_t[:] bounds,
+        np.int64_t n_neighbors=64, pbar=None, kernel_name="cubic",
+        use_normalization=True):
+
+    # hardcoding neighbors to test
+    cdef np.intp_t xsize, ysize, zsize
+    cdef np.float64_t w_j, coeff
+    cdef np.int64_t xi, yi, zi, pi
+    cdef np.float64_t q, h_j2, ih_j2
+    cdef int count, i, j, k, particle, pixel
+    cdef np.float64_t[:, :, :] buff_num
+    cdef np.float64_t[:, :, :] buff_denom
+    cdef np.float64_t[:,:,:,:] distances
+    cdef np.int64_t[:,:,:,:] pids
+    cdef np.int64_t[:] idx
+
+    xsize, ysize, zsize = buff.shape[0], buff.shape[1], buff.shape[2]
+    buff_num = np.zeros((xsize, ysize, zsize), dtype='f8')
+    buff_denom = np.zeros((xsize, ysize, zsize), dtype='f8')
+
+    kernel_func = get_kernel_func(kernel_name)
+
+    # generate the kdtree and find neighbours
+    # I think there is a better way of the passing than making a new array
+    kdtree = PyKDTree(np.array([posx, posy, posz]).T,
+                left_edge=np.array([bounds[0], bounds[2], bounds[4]]),
+                right_edge=np.array([bounds[1], bounds[3], bounds[5]]),
+                periodic=np.array([False, False, False]),
+                leafsize=n_neighbors)
```

Comment:
Why not use the global kdtree attached to the dataset instead of allocating a new one here?

```
In [1]: ds.parameter_filename
Out[1]: '/Users/goldbaum/Documents/test/GadgetDiskGalaxy/snapshot_200.hdf5'

In [2]: ds.index.kdtree
yt : [INFO     ] 2018-06-22 09:32:50,242 Allocating for 1.191e+07 particles
Initializing coarse index : 100%|████████████████| 19/19 [00:02<00:00, 10.81it/s]
Initializing refined index: 100%|████████████████| 19/19 [00:03<00:00,  6.02it/s]
yt : [INFO     ] 2018-06-22 09:32:57,243 Loading KDTree from snapshot_200.hdf5.kdtree
yt : [INFO     ] 2018-06-22 09:32:59,776 Detected hash mismatch, regenerating KDTree
yt : [INFO     ] 2018-06-22 09:32:59,873 Allocating KDTree for 4334546 particles
Out[2]: <cykdtree.kdtree.PyKDTree at 0x11f9af2c8>
```

See: https://github.com/yt-project/yt/blob/yt-4.0/yt/frontends/sph/data_structures.py#L125-L143

Reply:
ah, thank you very much! I didn't realise this existed